5-design-patterns/shipment/letter.ts

```diff
@@ -0,0 +1,26 @@
+import { IIdGenerator } from '../persistence/id-generator.model';
+import { Shipper } from '../shipper/Shipper';
+import { Shipment } from './Shipment';
+import { IShipment, IShipmentDTO, ShipmentType } from './shipment.model';
+
+export class Letter extends Shipment implements IShipment {
+  type: ShipmentType;
+
+  private constructor(shipment: IShipmentDTO, idGenerator?: IIdGenerator) {
+    super(shipment, idGenerator);
+    this.type = ShipmentType.LETTER;
+  }
+
+  static getInstance(
+    shipment: IShipmentDTO,
+    idGenerator?: IIdGenerator
+  ) {
+    return new Letter(shipment, idGenerator);
+  }
+  
+  ship(shipper: Shipper) {
+    const cost = shipper.getLetterCost(this.weight);
```

Comment:
```suggestion
    const cost = shipper.getCost(this.type, this.weight);
```
I'd say that's a tight coupling - calling a specific function for a `Letter`.
Having a general function, into which you'd be passing `type` and a `weight` looks like a better option.

Reply:
Fixed.